### PR TITLE
Fusion-managed data

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fusion-managed data is a new mechanism to communicate data between fusion and lowering that survives fusion clone.

To manage a data, use
```C++
size_t index = fusion.manage(data);
```
or
```C++
fusion.manage(key, data);
```

To get managed data, use
```C++
fusion.getManaged<T>(index);
```
or
```C++
fusion.getManaged<T>(key);
```

Current users of this mechanism are bank conflict info and loop rotation parameter.

### CODE REVIEW

File: csrc/fusion.h [hunk 1]
```diff
@@ -146,8 +148,11 @@ class TORCH_CUDA_CU_API Fusion : public IrContainer {
   bool isNoOp();
 
   //! Lower the fusion and evaluate bank conflict info
-  std::unordered_map<std::string, std::pair<int, int>> bankConflictInfo(
-      const CompileParams& compile_params = CompileParams());
+  //! Returns (tensor, read conflict ways, write conflict ways)
+  //! Each tensor can be read/write by multiple expressions, so the ways are
+  //! vectors.
+  std::unordered_map<TensorView*, std::pair<std::vector<int>, std::vector<int>>>
```
Comment (zasdfgbnm) [inline csrc/fusion.h:154]: Now I have a mechanism to associate bank conflict info with tensors in the fusion, so it no longer needs to return string as key anymore.

File: csrc/fusion.h [hunk 2]
```diff
@@ -310,9 +343,9 @@ class TORCH_CUDA_CU_API Fusion : public IrContainer {
   bool all_tv_uses_valid_ = false;
   bool is_during_update_uses_ = false;
 
-  // Compilation parameters guiding the loop rotation pass. See note
-  // [Loop Rotation] for detail.
-  LoopRotationParam loop_rotation_param_;
```
Comment (zasdfgbnm) [inline csrc/fusion.h:315]: Loop rotation parameter is now a managed data with key `loop_rotation`.

**Comment by xwang233** [comment]
!build

**Comment by xwang233** [comment]
!dontusethis

File: csrc/lower2device.h [hunk 3]
```diff
@@ -62,6 +62,10 @@ class TORCH_CUDA_CU_API GpuLower : public NonCopyable {
 
   kir::Kernel* kernel() const;
 
+  Fusion* fusion() const {
```
Comment (naoyam) [inline csrc/lower2device.h:65]: Looks like this just returns the same pointer as the above `kernel()`.

File: csrc/fusion.h [hunk 4]
```diff
@@ -242,21 +247,51 @@ class TORCH_CUDA_CU_API Fusion : public IrContainer {
...
-  // will be rotated
-  using LoopRotationParam = std::vector<
-      std::tuple<TensorView*, int64_t, std::unordered_set<Statement*>>>;
+  // Fusion-managed data is a mechanism to communicate data that survives fusion
+  // clone.
+  using CloneFn = std::function<std::any(IrCloner&, std::any)>;
 
-  const LoopRotationParam& getLoopRotationParam() const {
-    return loop_rotation_param_;
+  inline size_t manage(std::any data, CloneFn clone) {
```
Comment (naoyam) [inline csrc/fusion.h:254]: What is this `manage` supposed to do? Can you please add comments?

File: csrc/lower2device.h (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/lower2device.h:65]: I wasn't aware of that. Thanks for pointing out!

File: csrc/fusion.h (hunk 4 above)
Comment (zasdfgbnm) [reply csrc/fusion.h:306]: I added detailed doc and a test `FusionManagedData_CUDA` to explain this.

File: csrc/lower2device.h (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/lower2device.h]: fixed

**Comment by naoyam** [comment]
I wonder if we want to associate data to `IrContainer` rather than `Fusion`. There's no practical difference at this point, but `IrCloner` can be obtained from `IrContainer::copy` as well.

File: csrc/ir_cloner.h [hunk 5]
```diff
@@ -146,4 +146,20 @@ T* IrBuilder::clone(const T* src, IrCloner* ir_cloner) {
 template <typename T>
 NVFUSER_DEFINE_CLONE(Attribute<T>)
 
+template <typename T>
```
Comment (naoyam) [inline csrc/ir_cloner.h:149]: Why put these functions here and not in fusion.h?

Comment (zasdfgbnm) [reply csrc/ir_cloner.h:149]: Because `fusion.h` can not include `ir_cloner.h` (cyclic dependency), so I can not define them in `fusion.h`.

**Pull request merged**